lib/_stream_writable.js

```diff
@@ -302,7 +302,7 @@ function writeOrBuffer(stream, state, chunk, encoding, cb) {
 
   state.length += len;
 
-  var ret = state.length < state.highWaterMark;
+  var ret = state.length < state.highWaterMark && state.pendingcb < state.highWaterMark;
```

Comment:
we are comparing two completely different things here: `highWaterMark` is measured in objects or bytes, while `pendingcb` is measured in numbers of writes. I _think_ we should get a better fix for this problem, there is probably another counter we can check.